aten/src/ATen/native/cuda/ScanKernels.cu

```diff
@@ -347,6 +342,53 @@ __global__ void tensor_kernel_scan_innermost_dim(T *tgt_, T *src_,
   }
 }
 
+template <
+    typename T,
+    int num_threads_x,
+    int num_threads_y,
+    class BinaryFunction>
+__global__ typename std::enable_if<!c10::is_complex_t<T>::value, void>::type
+tensor_kernel_scan_innermost_dim(
+    T* tgt_,
+    T* src_,
+    unsigned num_rows,
+    unsigned row_size,
+    T init,
+    BinaryFunction binary_op) {
+  __shared__ T sbuf[num_threads_y][2 * num_threads_x];
+  T* row_buf = sbuf[threadIdx.y];
+
+  tensor_kernel_scan_innermost_dim_impl<T, num_threads_x, num_threads_y>(
+      row_buf, tgt_, src_, num_rows, row_size, init, binary_op);
+}
+
+template <
+    typename T,
+    int num_threads_x,
+    int num_threads_y,
+    class BinaryFunction>
+__global__ typename std::enable_if<c10::is_complex_t<T>::value, void>::type
+tensor_kernel_scan_innermost_dim(
+    T* tgt_,
+    T* src_,
+    unsigned num_rows,
+    unsigned row_size,
+    T init,
+    BinaryFunction binary_op) {
+  // As we cannot directly initialize shared array for complex types
+  // Reference:
+  //  `error: initializer not allowed for __shared__ variable`
+  // We instead get the base scalar type and allocate twice number of
+  // elements required of base type and reinterpret them as complex.
+  using base_t = typename scalar_value_type<T>::type;
+  __shared__ base_t sbuf[num_threads_y][4 * num_threads_x];
```

Comment:
is this effectively just viewing complex as float? If yes, it would work for `cumsum`, however it won't work for `cumprod` as is, since multiplication for two complex numbers is not defined as a dot product. `(a+bi)*(c+di) = ac-bd + (ad+bc)i`

Reply:
Here we are allocating shared memory buffer and not viewing into the actual data.

Since we can't allocate the shared memory for the `c10::complex` as it requires explicit initialization. So instead we allocate the equal memory in terms of the base scalar type and view that as a buffer for `c10::complex`.

So we are not actually viewing it as independent base type memory in the `tensor_kernel_scan_innermost_dim_impl` but as actual `c10::complex`. 
So cumprod won't be an issue.

Plus all the tests pass.

Reply:
Could you explain why `we can't allocate the shared memory for the c10::complex as it requires explicit initialization.`? `c10::complex` has a default constructor, and shouldn't it be initialized to its default value automatically?

Reply:
I am not quite sure why we can't allocate the shared memory for the `c10::complex` either. @kshitij12345 can you explain how is it related to explicit initialization requirement of `c10::complex`?

Reply:
A minimum repro of this problem could be
```C++
struct A {
    int a = 0;
    A() = default;
};

__global__ void f(A *p) {
    __shared__ A s[5];  // error: error: initializer not allowed for __shared__ variable 
    s[0] = *p;
}
```

Reply:
offline sync with @zasdfgbnm :
`__shared__ variables cannot have an initialization as part of their declaration` as explained here http://www.ncsa.illinois.edu/People/kindr/projects/hpca/files/singapore_p2.pdf

updating the complex constructor to not assign `real` and `imag` in constructor would resolve the problem.

Reply:
@anjali411 @zasdfgbnm Thanks. I was actually misguided in understanding the error. 

Thanks for the reference as well.